PyDSS/pyPostprocessor/PostprocessScripts/EdLiFoControl.py

```diff
@@ -0,0 +1,595 @@
+			
+	return  PVsys_dict, pd.DataFrame.from_dict(PVsys_dict, 'index')   
+	
+
+def get_monitored_line_dataframe(phase_info=3):
+	"""
+	Sample syntax:
+	monitored_lines = get_monitored_line_dataframe()
+	"""
+	lines = dss.utils.lines_to_dataframe()
+	column_list = [c.strip().lower() for c in lines.columns]
+	lines.columns = column_list
+	lines['phases'] = pd.to_numeric(lines['phases'])
+	if phase_info !=3:
+		monitored_lines = lines
+	else:
+		monitored_lines = lines.loc[lines['phases']==3,['bus1','bus2','rmatrix']].copy()
+		
+	return monitored_lines
+	
+def check_line_overloads(monitored_lines):
+	#monitored_lines = get_monitored_line_dataframe()
+	ovrl = None
+	
+	overloaded_line_dict = dict()
+	affected_buses = []
+	dss.Circuit.SetActiveClass("Line")
+	flag = dss.ActiveClass.First()
+	while flag>0:
+		line_name = dss.CktElement.Name()
+		line_limit = dss.CktElement.NormalAmps()
+		raw_current = dss.CktElement.Currents() 
+		line_current = [math.sqrt(i**2+j**2) for i,j in zip(raw_current[::2], raw_current[1::2])]
+		ldg = max(line_current)/float(line_limit)
+		#if ldg > 1.0 and dss.CktElement.NumPhases==3:
+		if ldg > 1.0:
+			#print('Checking Line: ', line_name)
+			#if line_name in monitored_lines.index:
+			overloaded_line_dict[line_name]= ldg*100
+			affected_buses.append(dss.CktElement.BusNames()[0])
+			affected_buses.append(dss.CktElement.BusNames()[1])
+			#affected_buses.append(monitored_lines.loc[line_name,'bus1'])
+			#affected_buses.append(monitored_lines.loc[line_name,'bus2'])
+		flag = dss.ActiveClass.Next()
+	
+	affected_buses = np.unique(list(affected_buses))
+	
+	if len(affected_buses)>0:
+		ovrl = pd.DataFrame.from_dict(overloaded_line_dict, 'index')
+		ovrl.columns = ['%normal']
+	
+	return ovrl, affected_buses
+
+def form_pvzones(affected_buses, dismat_df, zone_thr):
+	#print('Affected buses:', affected_buses)
+	ohmy = dismat_df.loc[affected_buses,:].copy()
+	zone=list()
+	for col in ohmy.columns:
+		if min(ohmy.loc[:,col])<=zone_thr:
+			zone.append(col)
+	
+	return zone
+    
+
+
+class EdLiFoControl(AbstractPostprocess):
+    """
+
+    """
+    REQUIRED_INPUT_FIELDS = ["curtailment_size",
+                             "electrical_distance_file_path",
+                             "zone_option",
+                             "zone_threshold",
+                             "fixed_pv_path",
+                             "lifo_pv_path",
+                             "lifo_min_pv_size",
+                             "user_lifo_pv_list" ]
+    # REQUIRED_INPUT_FIELDS = (
+        # "line_loading_limit",
+        # "dt_loading_limit",
+        # "line_safety_margin",
+        # "xfmr_safety_margin",
+        # "nominal_voltage",
+        # "max_iterations",
+        # "create_upgrade_plots",
+        # "tps_to_test",
+        # "create_upgrades_library",
+		# "upgrade_library_path",
+    # )
+
+    def __init__(self, project, scenario, inputs, dssInstance, dssSolver, dssObjects, dssObjectsByClass, simulationSettings, Logger):

```

Comment:
I suggest re-thinking how you are managing instance attributes. In many cases you are adding new attributes to `self` in methods other than `__init__` but then not using them in other methods. (self.Pct_curtailment, for example). Those could just be local variables.

It's a best practice to initialize all instance attributes in `__init__`.

Reply:
Okay

Reply:
Let's talk when you can